packages/@uppy/companion/src/server/provider/providerErrors.js

```diff
@@ -0,0 +1,40 @@
+const logger = require('../logger')
+const { ProviderApiError, ProviderAuthError } = require('./error')
+
+function convertProviderError ({ err, providerName, isAuthError = () => false, getJsonErrorMessage }) {
+  const { response } = err
+
+  function getErrorMessage () {
+    if (typeof response.body === 'object') {
+      const message = getJsonErrorMessage(response.body)
+      if (message != null) return message
+    }
+
+    if (typeof response.body === 'string') {
+      return response.body
+    }
+
+    return `request to ${providerName} returned ${response.statusCode}`
+  }
+
+  if (response) {
+    // @ts-ignore
+    if (isAuthError(response)) return new ProviderAuthError()
+
+    return new ProviderApiError(getErrorMessage(), response.statusCode)
+  }
+
+  return err
+}
+
+async function withProviderErrorHandling ({ fn, tag, providerName, isAuthError, getJsonErrorMessage }) {
+  try {
+    return await fn()
+  } catch (err) {
+    const err2 = convertProviderError({ err, providerName, isAuthError, getJsonErrorMessage })
+    logger.error(err2, tag)
+    throw err2
+  }
+}
```

Comment:
Interesting pattern!

Reply:
hah! probably not a textbook pattern but I wanted to reuse the error handling logic to make it consistent across all providers. maybe it could have been done with inheritance too

Reply:
alternatively we could pull the error handling logic out of the providers and instead do it in one place in the code that calls the provider methods

Reply:
Inheritance is rarely the answer for me. I think it's alright as your basically doing [inversion of control](https://kentcdodds.com/blog/inversion-of-control), and you're nudging towards going functional, as it's almost a `Maybe` monad, where you wrap a unpredictable action with flow control (stop or continue)